Clean/Gadgets/Addition8/Addition8FullCarry.lean

```diff
@@ -130,41 +128,52 @@ def circuit : FormalCircuit (F p) (Inputs p) (Outputs p) where
 
   completeness := by
    -- introductions
-    rintro ctx inputs inputs_var h_inputs
+    rintro i0 env inputs_var henv inputs h_inputs
     let ⟨x, y, carry_in⟩ := inputs
     let ⟨x_var, y_var, carry_in_var⟩ := inputs_var
     rintro as
 
     -- characterize inputs
-    have hx : x_var.eval = x := by injection h_inputs
-    have hy : y_var.eval = y := by injection h_inputs
-    have hcarry_in : carry_in_var.eval = carry_in := by injection h_inputs
+    have hx : x_var.eval env = x := by injection h_inputs
+    have hy : y_var.eval env = y := by injection h_inputs
+    have hcarry_in : carry_in_var.eval env = carry_in := by injection h_inputs
 
     -- simplify assumptions
     dsimp [assumptions] at as
 
     -- unfold goal, (re)introduce names for some of unfolded variables
-    dsimp
+    dsimp [Boolean.circuit, assert_bool]
     rw [hx, hy, hcarry_in]
-    let z := FieldUtils.mod_256 (x + y + carry_in)
-    let carry_out := FieldUtils.floordiv (x + y + carry_in) 256
-    rw [←(by rfl : z = FieldUtils.mod_256 (x + y + carry_in))]
-    rw [←(by rfl : carry_out = FieldUtils.floordiv (x + y + carry_in) 256)]
+    set z := env.get i0
+    set carry_out := env.get (i0 + 1)
+    let hz' : env.get i0 = z := rfl
+    let hcout' : env.get (i0 + 1) = carry_out := rfl
+
+    -- simplify local witnesses
+    have hz : z = mod_256 (x + y + carry_in) := by
+      have henv0 := henv (0 : Fin 2)
+      dsimp at henv0
+      rwa [hx, hy, hcarry_in, hz'] at henv0
+
+    have hcarry_out : carry_out = floordiv (x + y + carry_in) 256 := by
+      have henv1 := henv (1 : Fin 2)
+      dsimp at henv1
+      rwa [hx, hy, hcarry_in, hcout'] at henv1
 
     -- now it's just mathematics!
     guard_hyp as : x.val < 256 ∧ y.val < 256 ∧ (carry_in = 0 ∨ carry_in = 1)
 
     let goal_byte := ByteTable.contains (vec [z])
     let goal_bool := carry_out = 0 ∨ carry_out = 1
     let goal_add := x + y + carry_in + -1 * z + -1 * (carry_out * 256) = 0
-    show goal_byte ∧ (True ∧ goal_bool) ∧ goal_add
+    show ((True ∧ goal_byte) ∧ True ∧ goal_bool) ∧ goal_add
```

Comment:
TODO: where does the extra `True` come from?
(the change in parenthesis comes from iteration order for Lists vs our inductive type)